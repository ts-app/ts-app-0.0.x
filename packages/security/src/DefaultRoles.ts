export enum DefaultRoles {
  // support for self role...
  Administrator = 'Administrator',
  User = 'User'
}
