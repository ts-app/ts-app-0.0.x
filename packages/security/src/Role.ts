/**
 * This type specifies a name and unique ID for a Role.
 */
export interface Role {
  id: string
  name: string
}
