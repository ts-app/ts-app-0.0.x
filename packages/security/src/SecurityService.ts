import { Profile, User } from '@ts-app/common'
import { FindWithCursor } from '@ts-app/mongo'

export interface SecurityService {
  signUp (input: { email: string, password: string }): Promise<{ error?: string, user?: User }>

  loginWithEmailPassword (input: { email: string, password: string }): Promise<{ error?: string, userId?: string, accessToken?: string, refreshToken?: string }>

  user (id: string): Promise<{ error?: string, user?: User }>

  users (input: { q?: string, limit?: number, cursor?: string }): Promise<FindWithCursor<User>>

  removeUser (id: String): Promise<number>

  seedUsers (input?: { force?: boolean, userCount?: number }): Promise<{ error?: string }>

  updateProfile (input: { id: string, profile: object }): Promise<{ error?: string }>

  reset (): Promise<{ error?: string }>

  // --- TODO: not impl

  loginWithUserId (id: string): Promise<{ error?: string, user?: User }>

  setProfile<T extends Profile> (input: { id: string, profile: T }): Promise<void>

  getProfile<T extends Profile> (input: { id: string }): Promise<T>

}
