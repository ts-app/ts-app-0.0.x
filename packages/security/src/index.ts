export * from './DefaultRoles'
export * from './MongoRoleService'
export * from './MongoSecurityService'
export * from './Role'
export * from './RoleService'
export * from './securitySchemaDefinition'
export * from './SecurityService'
export * from './validators'
