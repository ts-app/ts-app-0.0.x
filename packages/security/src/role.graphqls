type Query {
    userIsInRoles(userId: String!, roles: [String]!, group: String): Boolean
    getGroupsForUser(userId: String, role: String): GetGroupsForUserPayload
    findRolesWithCursor(q: String, limit: Int, cursor: String): FindRolesWithCursorPayload
    userRoles(userId: String!, group: String): GetRolesForUserPayload
    getUsersInRoles(roles: [String]!, group: String, limit: Int, cursor: String): GetUsersInRolesPayload
}

type Mutation {
    addUsersToRoles(userIds: [String]!, roles: [String], group: String): NoPayload
    createRole(name: String!): CreateRolePayload
    removeRole(name: String!): NoPayload
    removeUsersFromRoles(userIds: [String]!, roles: [String]!, group: String): NoPayload
    removeUsersFromAllRoles(userIds: [String]!): NoPayload
}

type Role {
    id: String
    name: String
}

type CreateRolePayload {
    error: String
    id: String
}

type GetGroupsForUserPayload {
    error: String
    groups: [String]
}

type FindRolesWithCursorPayload {
    error: String
    cursor: String
    docs: [Role]
}

type GetRolesForUserPayload {
    error: String
    roles: [String]
}

type GetUsersInRolesPayload {
    error: String
    cursor: String
    docs: [User]
}
