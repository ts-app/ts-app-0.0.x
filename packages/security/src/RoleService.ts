import { Role } from './Role'
import { FindWithCursor } from '@ts-app/mongo'
import { User } from '@ts-app/common'

export const RoleServiceConstants = {
  GLOBAL: '_global_'
}

/**
 * Service to manage user access via roles and groups.
 *
 */
export interface RoleService {
  /**
   * Add users to specified roles, optionally specifying a group. If group name is not specified,
   * roles are assigned to RoleServiceConstants.GLOBAL group.
   *
   * @param {{userIds: (string | string[]); roles: (string | string[]); group?: string}} input
   * @return {Promise<{error?: string}>}
   */
  addUsersToRoles (input: {
    userIds: string | string[], roles: string | string[],
    group?: string
  }): Promise<{ error?: string }>

  /**
   * Create specified role name.
   *
   * @param {{name: string}} input
   * @return {Promise<{error?: string; id: string}>}
   */
  createRole (input: { name: string }): Promise<{ error?: string, id: string }>

  /**
   * Delete specified role name.
   *
   * If role is in use, service output error with appropriate message will be returned.
   *
   * @param {{name: string}} input
   * @return {Promise<{error?: string}>}
   */
  removeRole (input: { name: string }): Promise<{ error?: string }>

  /**
   * Get user's groups where user has role(s) assigned. RoleServiceConstants.GLOBAL will be omitted.
   *
   * If role is specified, restrict groups returned to those with specified role assigned.
   *
   * @param {{userId: string; role?: string}} input
   * @return {Promise<{error?: string; groups?: string[]}>}
   */
  getGroupsForUser (input: { userId: string, role?: string }): Promise<{ error?: string, groups?: string[] }>

  findRolesWithCursor (input?: {
    q?: string, limit?: number,
    cursor?: string
  }): Promise<FindWithCursor<Role>>

  /**
   * Get roles assigned to user, optionally restricting specifying a group. Roles assigned to
   * RoleServiceConstants.GLOBAL will be included.
   *
   * If group name is not specified, all roles including those assigned to
   * RoleServiceConstants.GLOBAL are returned.
   *
   * @param {string} userId
   * @param {string} group
   * @return {Promise<{error?: string; roles?: string[]}>}
   */
  getRolesForUser (input: { userId: string, group?: string }): Promise<{ error?: string, roles?: string[] }>

  getUsersInRoles (input: {
    roles: string | string[], group?: string, limit?: number,
    cursor?: string
  }): Promise<FindWithCursor<User>>

  removeUsersFromRoles (input: {
    userIds: string | string[], roles: string | string[],
    group?: string
  }): Promise<{ error?: string }>

  removeUsersFromAllRoles (input: { userIds: string | string[] }): Promise<{ error?: string }>

  /**
   * Returns true if user is assigned to one or more of the specified rule(s). This can be further
   * restricted based on group name.
   *
   * @param {{userId: string; roles: (string | string[]); group?: string}} input
   * @return {Promise<boolean>}
   */
  userIsInRoles (input: { userId: string, roles: string | string[], group?: string }): Promise<boolean>
}
