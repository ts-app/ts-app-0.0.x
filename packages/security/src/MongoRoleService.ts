import { RoleService, RoleServiceConstants } from './RoleService'
import { Role } from './Role'
import { ObjectId } from 'mongodb'
import { FindWithCursor, MongoService } from '@ts-app/mongo'
import { escapeRegex, User } from '@ts-app/common'
import { Resolver, ServiceInfo, userIdMatchParam } from '@ts-app/graphql'
import { DefaultRoles } from './DefaultRoles'

/**
 * Concept & implementation based on alanning/roles-npm
 */
export class MongoRoleService implements RoleService, ServiceInfo {
  private mongoService: MongoService

  constructor ({ mongoService }: { mongoService: MongoService }) {
    this.mongoService = mongoService
  }

  info (): { serviceName: string; } {
    return {
      serviceName: 'RoleService'
    }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async addUsersToRoles ({ userIds, roles, group = RoleServiceConstants.GLOBAL }: {
    userIds: string | string[], roles: string | string[], group?: string
  }): Promise<{ error?: string }> {
    if (!userIds || !roles || !group) {
      return { error: 'Invalid arguments for addUsersToRoles()' }
    }

    const _userIds = typeof userIds === 'string' ? [ userIds ] : userIds
    const _roles = typeof roles === 'string' ? [ roles ] : roles
    let _group = group

    if (_group) {
      if (_group.startsWith('$')) {
        return { error: 'Group cannot start with \'$\'' }
      }
      _group = _group.replace(/\./g, '_')
    }

    const rolesWithGroup = _roles.map(role => ({ role, group: _group }))

    // --- create missing roles in roles collection
    const cursor = await this.mongoService.cursor<Role>('roles', { name: { $in: _roles } })
    const matchingRolesInDb = await cursor.toArray()
    const rolesNotInDb = _roles.filter(roleName => !matchingRolesInDb.find(roleInDb => roleInDb.name === roleName))
    for (const roleName of rolesNotInDb) {
      const createRole = await this.createRole({ name: roleName })
      if (!createRole) {
        return { error: `Error creating role [${roleName}]` }
      }
    }

    await cursor.close()

    // --- assign roles to users collection
    const Users = await this.mongoService.collection('users')
    const updateMany = await Users.updateMany(
      {
        _id: { $in: _userIds.map(id => new ObjectId(id)) }
      },
      {
        $addToSet: {
          roles: {
            $each: rolesWithGroup
          }
        }
      }
    )

    if (updateMany.result.ok === 1) {
      return {}
    } else {
      return { error: 'Error adding users to roles' }
    }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async createRole ({ name }: { name: string }): Promise<{ error?: string, id: string }> {
    const id = await this.mongoService.create('roles', { name })
    return { id }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async removeRole ({ name }: { name: string }): Promise<{ error?: string }> {
    const userWithRole = await this.mongoService.get('users',
      {
        'roles.role': name
      },
      { fields: { _id: 1 } }
    )

    if (userWithRole) {
      return { error: `Role [${name}] is in use` }
    } else {
      await this.mongoService.remove('roles', { name })
      return {}
    }
  }

  @Resolver({ auth: userIdMatchParam('userId') })
  async getGroupsForUser ({ role, userId }: { userId: string, role?: string }): Promise<{ error?: string, groups?: string[] }> {
    let filter

    if (role) {
      filter = {
        _id: new ObjectId(userId),
        'roles.role': role
      }
    } else {
      filter = { _id: new ObjectId(userId) }
    }

    const user = await this.mongoService.get<User>('users', filter, {
      fields: {
        'roles': 1
      }
    })

    if (!user) {
      if (role) {
        return { error: `Error getting groups for user [${userId}] with role [${role}]` }
      } else {
        return { error: `Error getting groups for user [${userId}]` }
      }
    } else if (role) {
      const roles = user.roles || []
      const groupNames = roles.filter(currentRole => (currentRole.role === role && currentRole.group !== RoleServiceConstants.GLOBAL))
        .map(currentRole => {
          return currentRole.group
        })

      return { groups: Array.from(new Set(groupNames)) }
    } else {
      const roles = user.roles || []
      const groupNames = roles.filter(currentRole => (currentRole.group !== RoleServiceConstants.GLOBAL))
        .map(currentRole => currentRole.group)

      return { groups: Array.from(new Set(groupNames)) }
    }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ] })
  findRolesWithCursor (input: { q?: string, limit?: number, cursor?: string } = {}): Promise<FindWithCursor<Role>> {
    const { q, limit, cursor } = input
    let filter = {}
    if (q && q.trim().length > 0) {
      filter = {
        $or: [
          { name: { $regex: `^${escapeRegex(q)}`, $options: 'i' } },
          { name: { $regex: `${escapeRegex(q)}$`, $options: 'i' } }
        ]
      }
    }

    return this.mongoService.findWithCursor<Role>('roles', filter, limit, cursor)
  }

  @Resolver({ auth: userIdMatchParam('userId'), name: 'userRoles' })
  async getRolesForUser ({ userId, group }: { userId: string, group?: string }): Promise<{ error?: string, roles?: string[] }> {
    let filter
    if (!group) {
      filter = { _id: new ObjectId(userId) }
    } else {
      filter = {
        _id: new ObjectId(userId),
        'roles.group': { $in: [ RoleServiceConstants.GLOBAL, group ] }
      }
    }

    const user = await this.mongoService.get<User>('users', filter, {
      fields: {
        'roles': 1
      }
    })

    if (!user) {
      return { error: `Error getting roles for user [${userId}]` }
    } else if (user.roles) {
      // user with roles
      const roles = user.roles
        .filter(currentRole => {
          // filter for global groups
          // if groups is specified, filter by current role's group
          return currentRole.group === RoleServiceConstants.GLOBAL || !group || currentRole.group === group
        })
        .map(currentRole => currentRole.role)

      return { roles: Array.from(new Set(roles)) }
    } else {
      // user without roles
      return { roles: [] }
    }
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ] })
  async getUsersInRoles ({ group, roles, limit, cursor }: {
    roles: string | string[], group?: string, limit?: number, cursor?: string
  }): Promise<FindWithCursor<User>> {
    const _roles = typeof roles === 'string' ? [ roles ] : roles
    const _groups = group ? [ group, RoleServiceConstants.GLOBAL ] : [ RoleServiceConstants.GLOBAL ]

    let filter
    if (group) {
      filter = {
        roles: {
          $all: [
            { $elemMatch: { role: { $in: _roles }, group: { $in: _groups } } }
          ]
        }
      }
    } else {
      filter = {
        'roles.role': { $in: _roles }
      }
    }
    return this.mongoService.findWithCursor<User>('users', filter, limit, cursor)
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async removeUsersFromRoles ({ userIds, roles, group }: {
    userIds: string | string[], roles: string | string[], group?: string
  }): Promise<{ error?: string }> {
    const _userIds = typeof userIds === 'string' ? [ userIds ] : userIds
    const _roles = typeof roles === 'string' ? [ roles ] : roles
    const filter = {
      _id: { $in: _userIds.map(id => new ObjectId(id)) }
    }
    const update = {
      $pull: {
        roles: {
          role: { $in: _roles }
        }
      }
    }

    const collection = await this.mongoService.collection('users')
    const updateMany = await collection.updateMany(filter, update)

    if (updateMany.result.ok === 1) {
      return {}
    } else {
      return { error: `Error removing role(s) ${_roles} for user(s) ${_userIds}` }
    }

    // TODO: unused roles from collection('roles')
  }

  @Resolver({ auth: [ DefaultRoles.Administrator ], type: 'mutation' })
  async removeUsersFromAllRoles ({ userIds }: { userIds: string | string[] }): Promise<{ error?: string }> {
    const _userIds = typeof userIds === 'string' ? [ userIds ] : userIds
    const filter = {
      _id: { $in: _userIds.map(id => new ObjectId(id)) }
    }
    const update = {
      $unset: { roles: '' }
    }

    const collection = await this.mongoService.collection('users')
    const updateMany = await collection.updateMany(filter, update)

    if (updateMany.result.ok === 1) {
      return {}
    } else {
      return { error: `Error removing all roles for user(s) ${_userIds}` }
    }
  }

  @Resolver({ auth: userIdMatchParam('userId') })
  async userIsInRoles ({ userId, roles, group }: { userId: string, roles: string | string[], group?: string }): Promise<boolean> {
    try {
      const _role = typeof roles === 'string' ? [ roles ] : roles
      const _groups = group ? [ group, RoleServiceConstants.GLOBAL ] : [ RoleServiceConstants.GLOBAL ]

      const filter = {
        _id: new ObjectId(userId),
        roles: {
          $all: [
            { $elemMatch: { role: { $in: _role }, group: { $in: _groups } } }
          ]
        }
      }

      const findWithCursor = await this.mongoService.findWithCursor<User>('users', filter)
      return findWithCursor.docs.length > 0
    } catch (e) {
      // whenever error happens, return as user not in role
      console.error(e)
      return false
    }
  }
}
