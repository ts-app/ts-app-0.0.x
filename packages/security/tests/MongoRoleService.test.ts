import { MongoService } from '@ts-app/mongo'
import { SecurityService } from '../src/SecurityService'
import { RoleService } from '../src/RoleService'
import { MongoRoleService } from '../src/MongoRoleService'
import { MongoSecurityService } from '../src/MongoSecurityService'
import { User } from '@ts-app/common'
import arrayContaining = jasmine.arrayContaining

describe('MongoRoleService', async () => {
  const localUrl = 'mongodb://localhost:27017'
  let mongoService: MongoService
  let roleService: RoleService
  let securityService: SecurityService

  beforeEach(async () => {
    mongoService = new MongoService(localUrl)
    roleService = new MongoRoleService({ mongoService })
    securityService = new MongoSecurityService({ mongoService, roleService })

    try {
      await mongoService.dropCollection('users')
      await mongoService.dropCollection('roles')
    } catch {
      // it's ok
    }
  })

  afterEach(async () => {
    const db = await mongoService.db()
    return db.close()
  })

  /**
   * Create test fixtures. Tests in this file perform assertions based these test fixtures.
   *
   * WARNING: Changing these data WILL affect test assertions.
   *
   * THINK & ANALYZE carefully before changing this function.
   *
   */
  const createTestFixtures = async () => {
    const signUp = await securityService.signUp({ email: 'user1@test.com', password: '123' })
    expect(signUp.user!.id.length).toBe(24)
    const userId1 = signUp.user!.id

    const user2 = await securityService.signUp({ email: 'user2@test.com', password: '123' })
    expect(user2.user!.id.length).toBe(24)
    const userId2 = user2.user!.id

    const user3 = await securityService.signUp({ email: 'user3@test.com', password: '123' })
    expect(user3.user!.id.length).toBe(24)
    const userId3 = user3.user!.id

    expect((await roleService.addUsersToRoles({
      userIds: userId1, roles: [ 'user', 'superuser' ]
    })).error).toBeFalsy()

    expect((await roleService.addUsersToRoles({
      userIds: userId1, roles: [ 'user', 'superuser', 'admin' ], group: 'site 1'
    })).error).toBeFalsy()

    expect((await roleService.addUsersToRoles({
      userIds: userId1, roles: [ 'user', 'admin' ], group: 'site 2'
    })).error).toBeFalsy()
    expect((await roleService.addUsersToRoles({
      userIds: userId1, roles: [ 'user', 'superuser', 'admin' ], group: 'site 3'
    })).error).toBeFalsy()
    expect((await roleService.addUsersToRoles({
      userIds: userId1, roles: [ 'superuser' ], group: 'site 4'
    })).error).toBeFalsy()
    expect((await roleService.addUsersToRoles({
      userIds: userId2, roles: [ 'user', 'superuser', 'superadmin', 'god' ]
    })).error).toBeFalsy()
    expect((await roleService.addUsersToRoles({
      userIds: userId2, roles: [ 'user', 'superuser', 'admin' ], group: 'site 1'
    })).error).toBeFalsy()
    expect((await roleService.addUsersToRoles({
      userIds: userId2, roles: [ 'user', 'admin' ], group: 'site 2'
    })).error).toBeFalsy()
    expect((await roleService.addUsersToRoles({
      userIds: userId2, roles: [ 'user', 'superuser', 'admin' ], group: 'site 3'
    })).error).toBeFalsy()

    expect((await roleService.addUsersToRoles({
      userIds: userId3, roles: [ 'user' ]
    })).error).toBeFalsy()
    expect((await roleService.addUsersToRoles({
      userIds: userId3, roles: [ 'user', 'admin', 'superadmin' ], group: 'site 1'
    })).error).toBeFalsy()
    expect((await roleService.addUsersToRoles({
      userIds: userId3, roles: [ 'user' ], group: 'site 2'
    })).error).toBeFalsy()
    expect((await roleService.addUsersToRoles({
      userIds: userId3, roles: [ 'user', 'admin', 'god' ], group: 'site 3'
    })).error).toBeFalsy()

    return { userId1, userId2, userId3 }
  }

  test('addUsersToRoles()', async () => {
    const { userId1, userId2, userId3 } = await createTestFixtures()

    let user = await mongoService.get<User>('users', userId1)
    expect(user!.roles).toMatchSnapshot()

    user = await mongoService.get<User>('users', userId2)
    expect(user!.roles).toMatchSnapshot()

    user = await mongoService.get<User>('users', userId3)
    expect(user!.roles).toMatchSnapshot()
  })

  test('deleteRole()', async () => {
    await createTestFixtures()

    let removeRole = await roleService.removeRole({ name: 'superadmin' })
    expect(removeRole.error).toBe('Role [superadmin] is in use')

    const create = await roleService.createRole({ name: 'unused-role' })
    expect(create.error).toBeFalsy()
    expect(create.id.length).toBe(24)

    removeRole = await roleService.removeRole({ name: 'unused-role' })
    expect(removeRole.error).toBeFalsy()
  })

  test('getGroupsForUser()', async () => {
    const { userId1, userId2 } = await createTestFixtures()

    let groups = await roleService.getGroupsForUser({ userId: userId1 })
    expect(groups.error).toBeFalsy()
    expect(groups.groups!.length).toBe(4)
    expect(groups.groups).toEqual(arrayContaining([ 'site 1', 'site 2', 'site 3', 'site 4' ]))

    groups = await roleService.getGroupsForUser({ userId: userId2, role: 'superuser' })
    expect(groups.error).toBeFalsy()
    expect(groups.groups!.length).toBe(2)
    expect(groups.groups).toEqual(arrayContaining([ 'site 1', 'site 3' ]))
  })

  test('getGroupsForUser() with invalid role', async () => {
    // craete userId1
    const { userId1 } = await createTestFixtures()

    // userId1 does not have "superadmin" role
    const groups = await roleService.getGroupsForUser({ userId: userId1, role: 'superadmin' })
    expect(groups.error).toMatch(`Error getting groups for user [${userId1}] with role [superadmin]`)
  })

  test('getGroupsForUser() with invalid user', async () => {
    const userId = '012345678912'
    const groups = await roleService.getGroupsForUser({ userId })
    expect(groups.error).toBe(`Error getting groups for user [${userId}]`)
  })

  test('findRolesWithCursor()', async () => {
    await createTestFixtures()

    const roles = await roleService.findRolesWithCursor()
    expect(roles.cursor.length).toBe(24)

    const roleNames = roles.docs.map(role => role.name)
    expect(roleNames.length).toBe(5)
    expect(roleNames).toEqual(arrayContaining([ 'user', 'superuser', 'superadmin', 'admin', 'god' ]))
  })

  test('findRolesWithCursor() with query', async () => {
    await createTestFixtures()

    const roles = await roleService.findRolesWithCursor({ q: 'sU' })
    const roleNames = roles.docs.map(role => role.name)
    expect(roleNames.length).toBe(2)
    expect(roleNames).toEqual(arrayContaining([ 'superuser', 'superadmin' ]))
  })

  test('getRolesForUser()', async () => {
    const { userId1, userId2, userId3 } = await createTestFixtures()

    let user = await roleService.getRolesForUser({ userId: userId1 })
    expect(user.error).toBeFalsy()
    expect(user.roles!.length).toBe(3)
    expect(user.roles).toEqual(arrayContaining([ 'user', 'superuser', 'admin' ]))

    user = await roleService.getRolesForUser({ userId: userId2, group: 'unknown site' })
    expect(user.error).toBeFalsy()
    expect(user.roles).toEqual(arrayContaining([ 'user', 'superuser', 'superadmin' ]))

    user = await roleService.getRolesForUser({ userId: userId3, group: 'site 3' })
    expect(user.error).toBeFalsy()
    expect(user.roles!.length).toBe(3)
    expect(user.roles).toEqual(arrayContaining([ 'user', 'admin', 'god' ]))
  })

  test('getUsersInRole()', async () => {
    await createTestFixtures()

    let users = await roleService.getUsersInRoles({ roles: 'god' })
    let { docs } = users
    expect(docs.length).toBe(2)
    let emails = docs.map(user => user.emails[ 0 ].email)
    expect(emails).toEqual(arrayContaining([ 'user2@test.com', 'user3@test.com' ]))

    users = await roleService.getUsersInRoles({ roles: 'superadmin', group: 'site 1' })
    docs = users.docs
    expect(docs.length).toBe(2)
    emails = docs.map(user => user.emails[ 0 ].email)
    expect(emails).toEqual(arrayContaining([ 'user2@test.com', 'user3@test.com' ]))

    users = await roleService.getUsersInRoles({ roles: 'superadmin', group: 'site 2' })
    docs = users.docs
    expect(docs.length).toBe(1)
    emails = docs.map(user => user.emails[ 0 ].email)
    expect(emails).toEqual(arrayContaining([ 'user2@test.com' ]))

    users = await roleService.getUsersInRoles({ roles: 'superadmin', group: 'site x' })
    docs = users.docs
    expect(docs.length).toBe(1)
    emails = docs.map(user => user.emails[ 0 ].email)
    expect(emails).toEqual(arrayContaining([ 'user2@test.com' ]))
  })

  test('removeUsersFromRoles() - remove "def" from user1 but not user2', async () => {
    const { userId1, userId2 } = await createTestFixtures()

    let user = await roleService.getRolesForUser({ userId: userId1 })
    expect(user.roles!.length).toBe(3)
    expect(user.roles).toEqual(arrayContaining([ 'user', 'superuser', 'admin' ]))

    let removeUsersFromRoles = await roleService.removeUsersFromRoles({
      userIds: userId1, roles: 'user'
    })
    expect(removeUsersFromRoles.error).toBeFalsy()

    user = await roleService.getRolesForUser({ userId: userId1 })
    expect(user.roles!.length).toBe(2)
    expect(user.roles).toEqual(arrayContaining([ 'superuser', 'admin' ]))

    user = await roleService.getRolesForUser({ userId: userId2 })
    expect(user.roles!.length).toBe(5)
    expect(user.roles).toEqual(arrayContaining([ 'user', 'superuser', 'superadmin', 'god', 'admin' ]))

    removeUsersFromRoles = await roleService.removeUsersFromRoles({
      userIds: userId2, roles: [ 'user', 'god', 'admin', 'no match' ]
    })
    expect(removeUsersFromRoles.error).toBeFalsy()

    user = await roleService.getRolesForUser({ userId: userId2 })
    expect(user.roles!.length).toBe(2)
    expect(user.roles).toEqual(arrayContaining([ 'superuser', 'superadmin' ]))
  })

  test('removeUsersFromAllRoles()', async () => {
    const { userId1, userId2, userId3 } = await createTestFixtures()

    const removeUsersFromAllRoles = await roleService.removeUsersFromAllRoles({ userIds: [ userId1, userId2, userId3 ] })
    expect(removeUsersFromAllRoles.error).toBeFalsy()

    let user = await roleService.getRolesForUser({ userId: userId1 })
    expect(user.error).toBeFalsy()
    expect(user.roles!.length).toBe(0)

    user = await roleService.getRolesForUser({ userId: userId2 })
    expect(user.error).toBeFalsy()
    expect(user.roles!.length).toBe(0)

    user = await roleService.getRolesForUser({ userId: userId3 })
    expect(user.error).toBeFalsy()
    expect(user.roles!.length).toBe(0)
  })

  test('userIsInRoles()', async () => {
    const { userId1, userId2, userId3 } = await createTestFixtures()

    expect((await roleService.userIsInRoles(
      { userId: userId1, roles: 'superuser', group: 'site 2' }))).toBe(true)
    expect((await roleService.userIsInRoles(
      { userId: userId1, roles: 'admin', group: 'site 2' }))).toBe(true)
    expect((await roleService.userIsInRoles(
      { userId: userId1, roles: 'admin' }))).toBe(false)
    expect((await roleService.userIsInRoles(
      { userId: userId1, roles: 'admin', group: 'site 2' }))).toBe(true)

    expect((await roleService.userIsInRoles(
      { userId: userId2, roles: 'god' }))).toBe(true)
    expect((await roleService.userIsInRoles(
      { userId: userId2, roles: 'god', group: 'site 2' }))).toBe(true)
    expect((await roleService.userIsInRoles(
      { userId: userId2, roles: 'god', group: 'any site' }))).toBe(true)
    expect((await roleService.userIsInRoles(
      { userId: userId2, roles: 'admin' }))).toBe(false)
    expect((await roleService.userIsInRoles(
      { userId: userId2, roles: 'admin', group: 'any site' }))).toBe(false)
    expect((await roleService.userIsInRoles(
      { userId: userId2, roles: 'admin', group: 'site 2' }))).toBe(true)

    expect((await roleService.userIsInRoles(
      { userId: userId3, roles: 'admin' }))).toBe(false)
    expect((await roleService.userIsInRoles(
      { userId: userId3, roles: [ 'admin', 'god' ] }))).toBe(false)
    expect((await roleService.userIsInRoles(
      { userId: userId3, roles: [ 'admin', 'god' ], group: 'site 1' }))).toBe(true)
    expect((await roleService.userIsInRoles(
      { userId: userId3, roles: [ 'admin', 'god' ], group: 'site 2' }))).toBe(false)
    expect((await roleService.userIsInRoles(
      { userId: userId3, roles: [ 'admin', 'god' ], group: 'site 3' }))).toBe(true)
  })
})
