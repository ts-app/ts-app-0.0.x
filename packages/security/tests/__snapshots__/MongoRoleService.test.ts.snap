// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MongoRoleService addUsersToRoles() 1`] = `
Array [
  Object {
    "group": "_global_",
    "role": "user",
  },
  Object {
    "group": "_global_",
    "role": "superuser",
  },
  Object {
    "group": "site 1",
    "role": "user",
  },
  Object {
    "group": "site 1",
    "role": "superuser",
  },
  Object {
    "group": "site 1",
    "role": "admin",
  },
  Object {
    "group": "site 2",
    "role": "user",
  },
  Object {
    "group": "site 2",
    "role": "admin",
  },
  Object {
    "group": "site 3",
    "role": "user",
  },
  Object {
    "group": "site 3",
    "role": "superuser",
  },
  Object {
    "group": "site 3",
    "role": "admin",
  },
  Object {
    "group": "site 4",
    "role": "superuser",
  },
]
`;

exports[`MongoRoleService addUsersToRoles() 2`] = `
Array [
  Object {
    "group": "_global_",
    "role": "user",
  },
  Object {
    "group": "_global_",
    "role": "superuser",
  },
  Object {
    "group": "_global_",
    "role": "superadmin",
  },
  Object {
    "group": "_global_",
    "role": "god",
  },
  Object {
    "group": "site 1",
    "role": "user",
  },
  Object {
    "group": "site 1",
    "role": "superuser",
  },
  Object {
    "group": "site 1",
    "role": "admin",
  },
  Object {
    "group": "site 2",
    "role": "user",
  },
  Object {
    "group": "site 2",
    "role": "admin",
  },
  Object {
    "group": "site 3",
    "role": "user",
  },
  Object {
    "group": "site 3",
    "role": "superuser",
  },
  Object {
    "group": "site 3",
    "role": "admin",
  },
]
`;

exports[`MongoRoleService addUsersToRoles() 3`] = `
Array [
  Object {
    "group": "_global_",
    "role": "user",
  },
  Object {
    "group": "site 1",
    "role": "user",
  },
  Object {
    "group": "site 1",
    "role": "admin",
  },
  Object {
    "group": "site 1",
    "role": "superadmin",
  },
  Object {
    "group": "site 2",
    "role": "user",
  },
  Object {
    "group": "site 3",
    "role": "user",
  },
  Object {
    "group": "site 3",
    "role": "admin",
  },
  Object {
    "group": "site 3",
    "role": "god",
  },
]
`;
